File: src/main/java/com/mpe85/grampa/rule/impl/SequenceRule.java
package com.mpe85.grampa.rule.impl;

import static com.google.common.base.Preconditions.checkNotNull;

import java.util.List;

import com.mpe85.grampa.rule.Rule;
import com.mpe85.grampa.rule.RuleContext;

import one.util.streamex.StreamEx;

public class SequenceRule<T> extends AbstractRule<T> {
	
	public SequenceRule(final List<Rule<T>> rules) {
		super(checkNotNull(rules, "A list of 'rules' must not be null."));
	}
	
	@Override
	public boolean match(final RuleContext<T> context) {
		context.getStack().takeSnapshot();
		final boolean matched = StreamEx.of(getChildren()).allMatch(c -> context.createChildContext(c).run());
		context.getStack().removeSnapshot(!matched);
		return matched;
	}
	
}

File: src/main/kotlin/com/mpe85/grampa/rule/impl/SequenceRule.kt
package com.mpe85.grampa.rule.impl

import com.mpe85.grampa.context.ParserContext
import com.mpe85.grampa.rule.Rule
import com.mpe85.grampa.util.checkEquality
import com.mpe85.grampa.util.stringify
import java.util.Objects.hash

/**
 * A sequence rule implementation.
 *
 * @author mpe85
 * @param[T] The type of the stack elements
 * @param[rules] A list of child rules
 */
class SequenceRule<T>(rules: List<Rule<T>>) : AbstractRule<T>(rules) {

    override fun match(context: ParserContext<T>) = children.all { context.createChildContext(it).run() }

    override fun hashCode() = hash(super.hashCode())
    override fun equals(other: Any?) = checkEquality(other, { super.equals(other) })
    override fun toString() = stringify("#children" to children.size)
}

/**
 * Create a [SequenceRule] out of this and another rule.
 *
 * @param[other] Another rule
 * @return A [SequenceRule] with [this] and [other] as its child rules
 */
infix fun <T> Rule<T>.and(other: Rule<T>) = SequenceRule(listOf(this, other))

/**
 * Create a [SequenceRule] out of this and another rule.
 *
 * @param[other] Another rule
 * @return A [SequenceRule] with [this] and [other] as its child rules
 */
operator fun <T> Rule<T>.plus(other: Rule<T>) = this and other
